packages/compiler/src/core/checker.ts

```diff
@@ -2948,6 +2971,90 @@ export function createChecker(program: Program): Checker {
     }
   }
 
+  function checkObjectLiteral(node: ObjectLiteralNode, mapper: TypeMapper | undefined) {
+    return createAndFinishType({
+      kind: "ObjectLiteral",
+      node: node,
+      properties: checkObjectLiteralProperties(node, mapper),
+    });
+  }
+
+  function checkObjectLiteralProperties(
+    node: ObjectLiteralNode,
+    mapper: TypeMapper | undefined
+  ): Map<string, LiteralType | ObjectLiteral | TupleLiteral> {
+    const properties = new Map<string, LiteralType | ObjectLiteral | TupleLiteral>();
+
+    for (const prop of node.properties!) {
+      if ("id" in prop) {
+        const type = getTypeOrValueForNode(prop.value, mapper);
+        if (checkIsLiteralType(type, prop.value)) {
+          properties.set(prop.id.sv, type);
+        }
+      } else {
+        const targetType = checkObjectSpreadProperty(prop.target, mapper);
+        if (targetType) {
+          for (const [name, value] of targetType.properties) {
+            properties.set(name, value);
+          }
+        }
+      }
+    }
+    return properties;
+  }
+
+  function checkIsLiteralType(
+    type: Type,
+    diagnosticTarget: DiagnosticTarget
+  ): type is LiteralType | ObjectLiteral | TupleLiteral {
+    if (
+      type.kind !== "ObjectLiteral" &&
+      type.kind !== "TupleLiteral" &&
+      type.kind !== "String" &&
+      type.kind !== "Number" &&
+      type.kind !== "Boolean"
+    ) {
+      reportCheckerDiagnostic(createDiagnostic({ code: "not-literal", target: diagnosticTarget }));
+      return false;
+    }
+    return true;
+  }
+
+  function checkObjectSpreadProperty(
+    targetNode: TypeReferenceNode,
+    mapper: TypeMapper | undefined
+  ): ObjectLiteral | undefined {
+    const targetType = getTypeOrValueForNode(targetNode, mapper);
+
+    if (targetType.kind === "TemplateParameter" || isErrorType(targetType)) {
+      return undefined;
+    }
+    if (targetType.kind !== "ObjectLiteral") {
+      reportCheckerDiagnostic(createDiagnostic({ code: "spread-object", target: targetNode }));
+      return undefined;
+    }
+
+    return targetType;
+  }
+
+  function checkTupleLiteral(node: TupleLiteralNode, mapper: TypeMapper | undefined): TupleLiteral {
+    const values = node.values
+      .map((itemNode) => {
+        const type = getTypeOrValueForNode(itemNode, mapper);
+        if (checkIsLiteralType(type, itemNode)) {
+          return type;
+        } else {
+          return undefined; // TODO: do we want to omit this or include an error type?
```

Comment:
todo, what should happen in those cases?

We could omit the type which would allow further check to keep moving as long as where it was used didn't depend on the order. Or we make it the errorType and that fails